internal/datacoord/compaction_trigger.go

```diff
@@ -287,20 +295,43 @@ func (t *compactionTrigger) updateSegmentMaxSize(segments []*SegmentInfo) error
 		return err
 	}
 
+	isDiskANN := false
 	for _, indexInfo := range resp.IndexInfos {
 		indexParamsMap := funcutil.KeyValuePair2Map(indexInfo.IndexParams)
 		if indexType, ok := indexParamsMap["index_type"]; ok {
 			if indexType == indexparamcheck.IndexDISKANN {
-				diskSegmentMaxRows, err := t.estimateDiskSegmentMaxNumOfRows(collectionID)
+				// If index type is DiskANN, recalc segment max size here.
+				isDiskANN = true
+				newMaxRows, err := t.reCalcSegmentMaxNumOfRows(collectionID, true)
 				if err != nil {
 					return err
 				}
-				for _, segment := range segments {
-					segment.MaxRowNum = int64(diskSegmentMaxRows)
+				if len(segments) > 0 && int64(newMaxRows) != segments[0].GetMaxRowNum() {
+					log.Info("segment max rows recalculated for DiskANN collection",
+						zap.Int64("old max rows", segments[0].GetMaxRowNum()),
+						zap.Int64("new max rows", int64(newMaxRows)))
+					for _, segment := range segments {
+						segment.MaxRowNum = int64(newMaxRows)
+					}
 				}
 			}
 		}
 	}
+	// If index type is not DiskANN, recalc segment max size using default policy.
+	if !isDiskANN && !t.testingOnly {
```

Comment:
not sure why `testingOnly` is needed here

Reply:
In compaction unit tests, we are constantly setting up a lot of testing segments with customized current # of rows and customized max # of rows.

If max # of rows is recalculated in every test, it would make testing very hard :(